File: apps/cli/src/verify.ts
/**
 * `masax verify` — the regression test for the format code.
 *
 * Decodes every record of every dataset and checks each one consumes exactly
 * its declared payload length. That check is strong: it is what caught the
 * array count width, the short presence bitmap and the integer endianness,
 * each of which had already passed on most of the data.
 */
import chalk from "chalk";
import { Dataset, findVariants, parseDdm, parseFdt, resolveFile } from "@masax/lex";
import { NodeSource } from "@masax/lex/node";
import { datasetLocations, type DatasetLocation } from "@masax/catalogue";

export interface VerifyOptions {
  generation: 1 | 2;
  schema: boolean;
  only?: string;
}

interface Result {
  location: DatasetLocation;
  variant?: string;
  rows: number;
  clean: number;
  error?: string;
}

/**
 * Catalogue ids come from `CInfo`, which lists them authoritatively.
 *
 * They must not be discovered from the filenames: `catalog`'s template is
 * `@.bin`, which glob-matches every `.bin` in the directory. An earlier sweep
 * did exactly that and reported a format failure that was really `PBOOK.BIN`
 * and `DESC_GB.BIN` being counted as catalogues.
 */
async function catalogueIds(source: NodeSource, dir: string): Promise<string[]> {
  const cinfo = await Dataset.open(source, dir, "CInfo");
  const ids: string[] = [];
  for await (const { record } of cinfo.records()) {
    const id = record["A0"];
    if (typeof id === "string" && id) ids.push(id);
  }
  await cinfo.close();
  return ids;
}

async function variantsFor(
  source: NodeSource,
  location: DatasetLocation,
  catalogues: string[],
): Promise<(string | undefined)[]> {
  if (location.variant === "none") return [undefined];
  if (location.variant === "catalogue") return catalogues;
  const ddmPath = await resolveFile(source, location.dir, `${location.name}.ddm`);
  const ddm = parseDdm(await source.readFile(ddmPath));
  const fdtPath = await resolveFile(source, location.dir, ddm.fdt ?? `${location.name}.fdt`);
  const fdt = parseFdt(await source.readFile(fdtPath));
  return findVariants(source, location.dir, fdt.binTemplate);
}

export async function verify(root: string, options: VerifyOptions): Promise<number> {
  const source = new NodeSource(root);
  const locations = datasetLocations(options.generation).filter(
    (l) => !options.only || l.name.toLowerCase() === options.only.toLowerCase(),
  );

  const dataDir = `EPC/DATA${options.generation}`;
  let catalogues: string[] = [];
  try {
    catalogues = await catalogueIds(source, dataDir);
  } catch (error) {
    console.error(chalk.red(`cannot read ${dataDir}/CInfo: ${(error as Error).message}`));
    return 1;
  }
  console.log(chalk.dim(`${catalogues.length} catalogues listed in ${dataDir}/CInfo`));

  const results: Result[] = [];
  for (const location of locations) {
    let variants: (string | undefined)[];
    try {
      variants = await variantsFor(source, location, catalogues);
    } catch (error) {
      if (location.optional) continue;
      results.push({ location, rows: 0, clean: 0, error: (error as Error).message });
      continue;
    }

    let present = 0;
    let rows = 0;
    let clean = 0;
    let firstError: string | undefined;

    for (const variant of variants) {
      let dataset: Dataset;
      try {
        dataset = await Dataset.open(source, location.dir, location.name, variant);
      } catch {
        continue; // no data files for this variant
      }
      present++;
      if (options.schema) console.log(describeSchema(dataset));

      const sorted = dataset.index.checkSorted();
      if (!sorted.ok) {
        firstError ??= `${location.name}: .pnt is not sorted at entry ${sorted.at}`;
      }

      rows += dataset.count;
      try {
        for await (const _entry of dataset.records()) clean++;
      } catch (error) {
        firstError ??= (error as Error).message;
      }
      await dataset.close();
    }

    if (present === 0) {
      if (!location.optional) {
        results.push({ location, rows: 0, clean: 0, error: "no data files present" });
      }
      continue;
    }
    results.push({ location, rows, clean, error: firstError });
  }

  let totalRows = 0;
  let totalClean = 0;
  let failed = 0;
  for (const r of results) {
    totalRows += r.rows;
    totalClean += r.clean;
    const label = `${r.location.dir}/${r.location.name}`;
    if (r.error) {
      failed++;
      console.log(`${chalk.red("FAIL")} ${label.padEnd(26)} ${r.clean}/${r.rows}  ${r.error}`);
    } else {
      console.log(
        `${chalk.green("ok  ")} ${label.padEnd(26)} ${String(r.clean).padStart(8)}/${String(r.rows).padEnd(8)} clean`,
      );
    }
  }

  const pct = totalRows === 0 ? 0 : (100 * totalClean) / totalRows;
  const line = `${totalClean}/${totalRows} records decode exactly (${pct.toFixed(4)}%)`;
  console.log("");
  console.log(failed === 0 ? chalk.green(line) : chalk.red(line));
  return failed === 0 && totalClean === totalRows ? 0 : 1;
}

function describeSchema(dataset: Dataset): string {
  const head =
    `${dataset.name}${dataset.variant ? `[${dataset.variant}]` : ""}  ` +
    `rows=${dataset.count}  ${dataset.fdt.sparse ? "sparse" : "dense"}  ` +
    `keys=[${dataset.fdt.keys.join(", ")}] entry=${dataset.index.entrySize}B`;
  const lines = [chalk.bold(head)];
  const render = (field: (typeof dataset.fields)[number], depth: number) => {
    const kind =
      field.storage === 1
        ? "group"
        : field.storage === 2
          ? "array[u8]"
          : field.storage === 4
            ? "array[u16]"
            : "";
    lines.push(
      `  ${"  ".repeat(depth)}${field.code.padEnd(3)} ${kind.padEnd(10)} ` +
        `w=${String(field.width).padEnd(4)} ${field.name}`,
    );
    for (const child of field.children) {
      const named = dataset.fields.find((f) => f.code === child.code);
      if (named) render(named, depth + 1);
    }
  };
  for (const field of dataset.fields) {
    if (dataset.fdt.tree.some((t) => t.code === field.code)) render(field, 0);
  }
  return lines.join("\n");
}

File: re/tools/verify.py
#!/usr/bin/env python3
"""Decode every record of every ASA dataset and check it consumes exactly its
declared payload length. Usage: verify.py <path to ASA/M60>  [--schema]"""
import sys, os, glob, struct
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import lexdb


def main(root, show_schema=False):
    dirs = [d for d in ('EPC/DATA1', 'EPC/DATA1/A', 'EPC/DATA1/B', 'EPC/DATA2',
                        'EPC/DATA2/A', 'EPC/DATA2/B', 'PROG')
            if os.path.isdir(os.path.join(root, d))]
    grand_ok = grand_bad = 0
    for d in dirs:
        for ddm in sorted(glob.glob(os.path.join(root, d, '*.ddm'))):
            name = os.path.basename(ddm)[:-4]
            # Variant discovery must not need the data files to exist, so it
            # reads the .fdt template directly rather than via a Dataset.
            try:
                fdt = lexdb.parse_fdt(ddm[:-4] + '.fdt')
            except Exception as e:
                print(f'FDT  {d}/{name}: {e}')
                continue
            vs = [None]
            if '@' in fdt['bin']:
                pre, post = fdt['bin'].split('@')
                found = set()
                for f in os.listdir(os.path.join(root, d)):
                    fl, pl, sl = f.lower(), pre.lower(), post.lower()
                    if fl.startswith(pl) and fl.endswith(sl) and len(f) > len(pre) + len(post):
                        found.add(f[len(pre):len(f) - len(post)])
                # '@.bin' (catalog) matches every .bin in the directory, so drop
                # ids that are really another dataset's file.
                others = {os.path.basename(x)[:-4].lower()
                          for x in glob.glob(os.path.join(root, d, '*.ddm'))}
                vs = sorted(v for v in found
                            if v.lower() not in others
                            and not any(v.lower().startswith(t) for t in
                                        ('desc_', 'spn_')))
                vs = vs or [None]
            n_ok = n_bad = rows = nv = 0
            for v in vs:
                try:
                    ds = lexdb.Dataset(ddm, v)
                except FileNotFoundError:
                    continue
                nv += 1
                rows += ds.count()
                if show_schema:
                    print(ds.schema()); print()
                data = open(ds.binpath, 'rb').read()
                for key, off in ds.index():
                    try:
                        ln, = struct.unpack_from('<H', data, off)
                        rec, used = ds.decode(data[off + 2:off + 2 + ln])
                        if used == ln: n_ok += 1
                        else: n_bad += 1
                    except Exception:
                        n_bad += 1
            if not nv:
                print(f'---- {d}/{name}: no data files present')
                continue
            grand_ok += n_ok; grand_bad += n_bad
            flag = 'OK  ' if n_bad == 0 else 'FAIL'
            print(f'{flag} {d}/{name:<12} {nv:>2} variant(s)  {n_ok:>7}/{rows:<7} clean'
                  + ('' if not n_bad else f'  {n_bad} BAD'))
    tot = grand_ok + grand_bad
    print(f'\n{grand_ok}/{tot} records decode exactly'
          f' ({100 * grand_ok / max(tot, 1):.4f}%)')
    return 0 if grand_bad == 0 else 1


if __name__ == '__main__':
    args = [a for a in sys.argv[1:] if not a.startswith('--')]
    sys.exit(main(args[0], '--schema' in sys.argv))
